src/Ethereum/ABI/Array.cpp

```diff
@@ -133,4 +133,44 @@ std::string ParamArray::getExtraTypes(std::vector<std::string>& ignoreList) cons
     return (proto != nullptr) ? proto->getExtraTypes(ignoreList) : "";
 }
 
-} // namespace TW::Ethereum::ABI
+void ParamArrayFix::encode(Data& data) const {
+    this->_params.encode(data);
+}
+
+bool ParamArrayFix::decode(const Data& encoded, size_t& offset_inout) {
+    std::size_t origOffset = offset_inout;
+    auto res = this->_params.decode(encoded, offset_inout);
+    // padding
+    offset_inout = origOffset + ValueEncoder::paddedTo32(offset_inout - origOffset);
+    return res;
+}
+
+bool ParamArrayFix::setValueJson(const std::string& value) {
+    auto valuesJson = json::parse(value, nullptr, false);
+    if (valuesJson.is_discarded() || !valuesJson.is_array() || _params.getCount() < valuesJson.size()) {
+        return false;
+    }
+
+    std::size_t idx{0};
+    for (auto&& e : valuesJson) {
+        std::string eString = e.is_string() ? e.get<std::string>() : e.dump();
+        _params.getParamUnsafe(idx)->setValueJson(eString);
+        ++idx;
+    }
+    return true;
+}
+
+void ParamArrayFix::addParams(const Params& params) {
+    auto addParamFunctor = [this](auto&& param) {
+        if (param == nullptr) {
+            throw std::runtime_error("param can't be nullptr");
```

Comment:
Exceptions are not working on wasm, maybe use assert(false && "message") so message won't be lost?

Reply:
Assertion break the coverage unit tests, I'm not against it but we will end up having preprocessor conditions to check the platform then.

Reply:
In general, asserts should be used to test for internal inconsistencies, but not for invalid input data.